File: src/loopflow/lf/naming.py
"""Branch naming utilities for loopflow.

Generates branch names from wave name + config schema.
"""

import random
import re
import subprocess
from datetime import datetime
from pathlib import Path

from loopflow.lf.config import load_config

# Word lists for generating unique branch names

MAGICAL = [
    "aurora",
    "cascade",
    "crystal",
    "drift",
    "echo",
    "ember",
    "fern",
    "flume",
    "frost",
    "glade",
    "grove",
    "haze",
    "ivy",
    "jade",
    "luna",
    "mist",
    "nova",
    "opal",
    "petal",
    "prism",
    "rain",
    "ripple",
    "sage",
    "shade",
    "spark",
    "star",
    "stone",
    "storm",
    "tide",
    "vale",
    "wave",
    "wisp",
    "wren",
    "zephyr",
]

MUSICAL = [
    "allegro",
    "aria",
    "ballad",
    "cadence",
    "canon",
    "chord",
    "coda",
    "duet",
    "forte",
    "fugue",
    "harmony",
    "hymn",
    "lilt",
    "lyric",
    "melody",
    "motif",
    "opus",
    "prelude",
    "refrain",
    "rondo",
    "sonata",
    "tempo",
    "trill",
    "tune",
    "verse",
    "waltz",
]


def generate_word_pair() -> str:
    """Generate a random magical-musical pair like 'aurora-melody'."""
    magical = random.choice(MAGICAL)
    musical = random.choice(MUSICAL)
    return f"{magical}-{musical}"


def generate_timestamp() -> str:
    """Generate timestamp in YYYYMMDD_HHMM format."""
    return datetime.now().strftime("%Y%m%d_%H%M")


def branch_exists(repo: Path, branch: str) -> bool:
    """Check if a branch exists locally or on origin."""
    result = subprocess.run(
        ["git", "rev-parse", "--verify", f"refs/heads/{branch}"],
        cwd=repo,
        capture_output=True,
    )
    if result.returncode == 0:
        return True
    result = subprocess.run(
        ["git", "rev-parse", "--verify", f"refs/remotes/origin/{branch}"],
        cwd=repo,
        capture_output=True,
    )
    return result.returncode == 0


def _is_timestamp(s: str) -> bool:
    """Check if string matches YYYYMMDD_HHMM format."""
    return bool(re.match(r"^\d{8}_\d{4}$", s))


def _is_word_pair(s: str) -> bool:
    """Check if string is a magical-musical word pair."""
    if "-" not in s:
        return False
    parts = s.split("-", 1)
    if len(parts) != 2:
        return False
    return parts[0] in MAGICAL and parts[1] in MUSICAL


def _get_git_username() -> str:
    """Get username from git config user.name or $USER env var."""
    import os

    result = subprocess.run(
        ["git", "config", "user.name"],
        capture_output=True,
        text=True,
    )
    if result.returncode == 0 and result.stdout.strip():
        return _sanitize_for_branch(result.stdout.strip())

    return _sanitize_for_branch(os.environ.get("USER", "user"))


def _sanitize_for_branch(s: str) -> str:
    """Replace spaces, special chars with hyphens for valid git branch names."""
    result = re.sub(r"[\s@#$%^&*()+=\[\]{}|\\:;\"'<>,?!~`]+", "-", s)
    result = result.strip("-")
    result = re.sub(r"-+", "-", result)
    return result.lower()


def parse_branch_base(branch: str, repo: Path) -> str:
    """Extract wave name from branch using the schema to know what to strip.

    Inverse of generate_branch_name. Uses the config schema to identify
    which parts are {user}, {ts}/{timestamp}, {words} and strips them,
    leaving only the {name} portion.

    Gets user from git config (same source as branch generation).

    Examples (with schema="{user}.{name}.{ts}", user="jack-heart"):
        'jack-heart.concerto.20260202_1700' → 'concerto'
        'jack-heart.concerto' → 'concerto'
        'concerto.20260202_1700' → 'concerto'
        'concerto' → 'concerto'
    """
    config = load_config(repo)

    if config and config.branch_names:
        schema = config.branch_names.schema_
    else:
        schema = "{user}.{name}.{timestamp}.{words}"

    # Get user from git config (same source as branch generation)
    user = _get_git_username()

    # Build a regex from the schema to extract {name}
    pattern = re.escape(schema)

    # {user} matches the git username
    pattern = pattern.replace(r"\{user\}", re.escape(user))

    # {name} is what we want to capture
    pattern = pattern.replace(r"\{name\}", r"(?P<name>.+?)")

    # {ts} and {timestamp} match YYYYMMDD_HHMM
    pattern = pattern.replace(r"\{ts\}", r"\d{8}_\d{4}")
    pattern = pattern.replace(r"\{timestamp\}", r"\d{8}_\d{4}")

    # {date} matches YYYYMMDD
    pattern = pattern.replace(r"\{date\}", r"\d{8}")

    # {words} matches magical-musical word pairs
    word_pattern = r"(?:" + "|".join(MAGICAL) + r")-(?:" + "|".join(MUSICAL) + r")"
    pattern = pattern.replace(r"\{words\}", word_pattern)

    # Try matching with full pattern first
    match = re.fullmatch(pattern, branch)
    if match and "name" in match.groupdict():
        return match.group("name")

    # Fallback: strip known suffixes manually
    parts = branch.split(".")

    # Strip trailing word pair and timestamp
    if len(parts) >= 2 and _is_word_pair(parts[-1]) and _is_timestamp(parts[-2]):
        parts = parts[:-2]
    elif len(parts) >= 1 and _is_timestamp(parts[-1]):
        parts = parts[:-1]

    # Strip leading user prefix
    if parts and parts[0] == user:
        parts = parts[1:]

    return ".".join(parts) if parts else branch


def generate_branch_name(wave_name: str, repo: Path) -> str:
    """Generate unique branch name for a wave using config schema.

    Schema placeholders:
        {user} - user identifier from config
        {name} - wave name
        {timestamp} - YYYYMMDD_HHMM format
        {words} - magical-musical word pair

    Default schema: "{user}.{name}.{timestamp}.{words}"
    Example: "jack-heart.concerto.20260202_1700.aurora-melody"
    """
    config = load_config(repo)

    # Get schema from config, or use default
    if config and config.branch_names:
        schema = config.branch_names.schema_
    else:
        schema = "{user}.{name}.{timestamp}.{words}"

    # Get user from config
    user = config.user if config else None
    if not user and "{user}" in schema:
        raise ValueError("Config missing 'user' field required by branch_names.schema")

    timestamp = generate_timestamp()

    for _ in range(100):
        words = generate_word_pair()
        candidate = schema.format(
            user=user or "",
            name=wave_name,
            timestamp=timestamp,
            words=words,
        )
        # Clean up any leading/trailing dots from empty placeholders
        candidate = candidate.strip(".")
        candidate = re.sub(r"\.+", ".", candidate)

        if not branch_exists(repo, candidate):
            return candidate

    raise ValueError(f"Could not generate unique branch for wave {wave_name}")

File: rust/loopflow-engine/src/naming.rs
use crate::config::BranchNameConfig;
use crate::error::GitError;
use chrono::Local;
use rand::seq::SliceRandom;
use rand::Rng;
use std::path::Path;
use std::process::Command;

const MAGICAL: &[&str] = &[
    "aurora", "cascade", "crystal", "drift", "echo", "ember", "fern", "flume", "frost", "glade",
    "grove", "haze", "ivy", "jade", "luna", "mist", "nova", "opal", "petal", "prism", "rain",
    "ripple", "sage", "shade", "spark", "star", "stone", "storm", "tide", "vale", "wave", "wisp",
    "wren", "zephyr",
];

const MUSICAL: &[&str] = &[
    "allegro", "aria", "ballad", "cadence", "canon", "chord", "coda", "duet", "forte", "fugue",
    "harmony", "hymn", "lilt", "lyric", "melody", "motif", "opus", "prelude", "refrain", "rondo",
    "sonata", "tempo", "trill", "tune", "verse", "waltz",
];

fn sanitize_for_branch(value: &str) -> String {
    let mut out = String::new();
    let mut last_was_dash = false;
    for ch in value.chars() {
        let lowered = ch.to_ascii_lowercase();
        let keep =
            lowered.is_ascii_alphanumeric() || lowered == '-' || lowered == '_' || lowered == '.';
        if keep {
            out.push(lowered);
            last_was_dash = false;
        } else if !last_was_dash {
            out.push('-');
            last_was_dash = true;
        }
    }
    let trimmed = out.trim_matches('-').to_string();
    let mut collapsed = String::new();
    let mut prev_dash = false;
    for ch in trimmed.chars() {
        if ch == '-' {
            if !prev_dash {
                collapsed.push(ch);
            }
            prev_dash = true;
        } else {
            collapsed.push(ch);
            prev_dash = false;
        }
    }
    if collapsed.is_empty() {
        "user".to_string()
    } else {
        collapsed
    }
}

fn git_username(repo: &Path) -> Result<String, GitError> {
    let output = Command::new("git")
        .arg("-C")
        .arg(repo)
        .args(["config", "user.name"])
        .output()?;
    if output.status.success() {
        let name = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if !name.is_empty() {
            return Ok(sanitize_for_branch(&name));
        }
    }
    let fallback = std::env::var("USER").unwrap_or_else(|_| "user".to_string());
    Ok(sanitize_for_branch(&fallback))
}

fn generate_word_pair_with_rng<R: Rng + ?Sized>(rng: &mut R) -> String {
    let magical = MAGICAL.choose(rng).copied().unwrap_or("wisp");
    let musical = MUSICAL.choose(rng).copied().unwrap_or("forte");
    format!("{magical}-{musical}")
}

fn generate_word_pair() -> String {
    let mut rng = rand::thread_rng();
    generate_word_pair_with_rng(&mut rng)
}

fn generate_timestamp() -> String {
    Local::now().format("%Y%m%d_%H%M").to_string()
}

fn generate_date() -> String {
    Local::now().format("%Y%m%d").to_string()
}

pub fn format_branch_name(
    short_name: &str,
    config: Option<&BranchNameConfig>,
    repo: &Path,
) -> Result<String, GitError> {
    let Some(config) = config else {
        return Ok(short_name.to_string());
    };

    let schema = config.schema_.as_str();
    if schema == "{name}" {
        return Ok(short_name.to_string());
    }

    let user = git_username(repo)?;
    let ts = generate_timestamp();
    let date = generate_date();
    let words = generate_word_pair();

    let mut result = schema.replace("{name}", short_name);
    result = result.replace("{user}", &user);
    result = result.replace("{timestamp}", &ts);
    result = result.replace("{ts}", &ts);
    result = result.replace("{date}", &date);
    result = result.replace("{words}", &words);

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::rngs::StdRng;
    use rand::SeedableRng;
    use tempfile::tempdir;

    #[test]
    fn sanitize_for_branch_cleans_input() {
        let cleaned = sanitize_for_branch("Jack Heart!!!");
        assert_eq!(cleaned, "jack-heart");
    }

    #[test]
    fn format_branch_name_passthrough_without_config() {
        let repo = tempdir().expect("tempdir");
        let name = format_branch_name("feature", None, repo.path()).expect("format");
        assert_eq!(name, "feature");
    }

    #[test]
    fn generate_word_pair_is_deterministic_with_rng() {
        let mut rng = StdRng::seed_from_u64(42);
        let pair = generate_word_pair_with_rng(&mut rng);
        assert!(!pair.is_empty());
        assert!(pair.contains('-'));
    }
}
